apis/projectcontour/v1alpha1/contourconfig.go

```diff
@@ -22,6 +25,396 @@ import (
 // It contains most of all the options that can be customized, the
 // other remaining options being command line flags.
 type ContourConfigurationSpec struct {
+	// XDSServer contains parameters for the xDS server.
+	// +optional
+	XDSServer ServerParameters `json:"xdsServer,omitempty"`
+
+	// IngressOptions contains parameters for ingress options.
+	// +optional
+	IngressOptions Ingress `json:"ingress,omitempty"`
+
+	EnvoyConfig Envoy `json:"envoy,omitempty"`
+
+	// GatewayConfig contains parameters for the gateway-api Gateway that Contour
+	// is configured to serve traffic.
+	// +optional
+	GatewayConfig *GatewayParameters `json:"gateway,omitempty"`
+
+	// HTTPProxy defines parameters on HTTPProxy.
+	// +optional
+	HTTPProxy *HTTPProxyConfig `json:"httpproxy,omitempty"`
+
+	// LeaderElection contains leader election parameters.
+	// +optional
+	LeaderElection LeaderElectionParameters `json:"leaderelection,omitempty"`
+
+	// EnableExternalNameService allows processing of ExternalNameServices
+	// Defaults to disabled for security reasons.
+	// +optional
+	EnableExternalNameService bool `json:"enableExternalNameService,omitempty"`
+
+	// RateLimitService optionally holds properties of the Rate Limit Service
+	// to be used for global rate limiting.
+	RateLimitService RateLimitService `json:"rateLimitService,omitempty"`
+
+	// Policy specifies default policy applied if not overridden by the user
+	Policy PolicyParameters `json:"policy,omitempty"`
+}
+
+// TLS holds TLS file config details.
+type TLS struct {
+	// CA filename.
+	CAFile string `json:"caFile,omitempty"`
+
+	// Client certificate filename.
+	CertFile string `json:"certFile,omitempty"`
+
+	// Client key filename.
+	KeyFile string `json:"keyFile,omitempty"`
+}
+
+// Ingress defines ingress specific config items.
+type Ingress struct {
+	// Ingress Class Name Contour should use.
+	// +optional
+	ClassName string `json:"className,omitempty"`
+
+	// Address to set in Ingress object status.
+	// +optional
+	StatusAddress string `json:"statusAddress,omitempty"`
+}
+
+// Health defines the endpoints Contour will serve to enable health checks.
+type Health struct {
+	// Defines the Contour health address interface.
+	//  +optional
+	// +kubebuilder:default:"0.0.0.0"
+	// +kubebuilder:validation:MinLength=1
+	Address string `json:"address,omitempty"`
+
+	// Defines the Contour health port.
+	// Defaults to 8000.
+	//  +optional
+	// +kubebuilder:default=8000
+	Port int `json:"port,omitempty"`
+}
+
+// Metrics defines the endpoints for metrics.
+type Metrics struct {
+	// Defines the Contour metrics address interface.
+	//  +optional
+	// +kubebuilder:validation:MinLength=1
+	Address string `json:"address,omitempty"`
+
+	// Defines the Contour metrics port.
+	// Defaults to 8000.
+	//  +optional
+	Port int `json:"port,omitempty"`
```

Comment:
can set a kubebuilder default here

Reply:
To do that we'll need to make different structs for each type as Envoy's metrics are different defaults than Contour's.

Reply:
Ah. I think you can also set a default at the place where the struct is used as a field - see e.g. https://github.com/projectcontour/contour-operator/blob/main/api/v1alpha1/contour_types.go#L73-L74